Cl/contact infos

LozanoCyrielle opened the pull request:
Add contact infos for users.
What I've done : 
- Add fields in user (email, phone number, telegram, favorite contact methods)
- At first all user have they're email fill, and the email is the favorite method by default
- Add those into the edit account
- Show those info when a loan is accepted or ongoing

Note that the item UI is not done for now (reject button and not cancel, but Massimo is working on it at the moment #289, so to avoid very painful merge conflict I used his branch)

Patrick-Pataky commented on app/src/main/java/com/android/partagix/ui/screens/EditAccount.kt:
```diff
@@ -157,20 +181,111 @@ fun EditAccount(
+                  TextField(
+                      modifier = Modifier.fillMaxWidth().padding(16.dp, 0.dp).testTag("email"),
+                      value = email,
+                      onValueChange = { email = it },
```
Here (and for the 2 other fields too), it would be great that if the user erases the entire input field, than the checkbox is unchecked and removed from `favorite[]`
```suggestion
                      onValueChange = { 
                         email = it 
                         if (it.isEmpty()) { favorite[0] = !favorite[0] }
                      },
```

Patrick-Pataky commented on app/src/main/java/com/android/partagix/ui/screens/EditAccount.kt:
```diff
@@ -157,20 +181,111 @@ fun EditAccount(
+                        Row(
+                            horizontalArrangement = Arrangement.Absolute.Left,
+                            verticalAlignment = Alignment.CenterVertically) {
+                              Checkbox(
```
One another cool thing with those Checkboxes would be to hide the Input Field if the corresponding Checkbox is unchecked

LozanoCyrielle replied:
For me we still need to be able to enter other contact information, for example my favorite is telegram but I know everybody doesn't have telegram so I also put my phone number

Patrick-Pataky replied:
OH I see, I misunderstand it. Than it works

Patrick-Pataky reviewed (commented):
Globally, it is a really nice PR. 
Additional open question:
- Do we want to add contact information on the ManageLoanRequest screen ?
- For telegram for example, do we want to add an intent that launches telegram and opens a new conversion with the user, when we click on the username ?

LozanoCyrielle commented on app/src/main/java/com/android/partagix/ui/components/RankingStars.kt:
```diff
@@ -0,0 +1,110 @@
+ * @return a string with the corresponding amount of stars
+ */
+@Composable
+fun RankStars(rank: String) {
```
it's not mine but Massimo's, I merge his branch to avoid big merge conflict later

LozanoCyrielle commented:
> Globally, it is a really nice PR. Additional open question:
> 
> * Do we want to add contact information on the ManageLoanRequest screen ?
> * For telegram for example, do we want to add an intent that launches telegram and opens a new conversion with the user, when we click on the username ?

> I put he infos only when the loan is accepted or ongoing, it can appear in all screens as long as the list is set to expendable. I put the infos only for accepted and ongoing because people might not want to put their infos if they are always visible
> I can try using intent but maybe more like a copy the username/number/email to clipboard when we clic ? so the intent is valid for all of them. But I'm not sure I will be able to do it

LozanoCyrielle commented:
I added intents:
- opening telegram (or the website)
- opening the messaging app
- opening email creation

Also sometimes when expanding the loan in inventory the app was crashing I found out why I fixed it -> but now we cannot click directly as the crash was caused by something that wasn't loaded

I changed the UI in edit screen to try an make it more clear for a user experience and the checkbox uncheck when the field is empty

Patrick-Pataky commented on app/src/main/java/com/android/partagix/ui/screens/EditAccount.kt:
```diff
@@ -157,20 +181,126 @@ fun EditAccount(
+
+                  TextField(
+                      modifier = Modifier.fillMaxWidth().padding(16.dp, 0.dp).testTag("email"),
+                      value = email,
```
The entered information should fit on a single line. The same modification can be applied to the 2 other fileds.
```suggestion
                      value = email,
                      singleLine = true,
```

LozanoCyrielle replied:
yes you're right !

Patrick-Pataky commented on app/src/main/java/com/android/partagix/ui/components/ItemUi.kt:
```diff
@@ -44,264 +48,469 @@ import androidx.compose.ui.text.style.TextOverflow
+      fontWeight = FontWeight.Bold,
+      modifier =
+          Modifier.clickable {
+            val intent =
```
Those Intents are really value to the app, having much more interactivity overall. Great job !

Patrick-Pataky commented on app/src/main/java/com/android/partagix/ui/components/ItemUi.kt:
```diff
@@ -44,264 +48,469 @@ import androidx.compose.ui.text.style.TextOverflow
+      }
+    }
+  }
+  /*} else {
```
Maybe this commented code will be gone with the merge with the main, but make sure it won't be visible at the end

Patrick-Pataky commented on app/src/main/java/com/android/partagix/ui/components/ItemUi.kt:
```diff
@@ -44,264 +48,469 @@ import androidx.compose.ui.text.style.TextOverflow
+          Modifier.clickable {
+            val intent =
+                when {
+                  contact.startsWith("Email :") ->
```
Sorry to bother with the small details, to make it clean, the best would be to create an ENUM with those constant values ("Email :", "Phone :", "Telegram :"). 
You can take inspiration by looking at the Visibility ENUM.

LozanoCyrielle replied:
I think that using an enum as it's the only part of the code where I iterate on this value will just make the code more complicated to read, and won't add much value

Patrick-Pataky reviewed (changes requested):
Overall, really nice changes ! Continue like that, thank you

LozanoCyrielle commented:
Definitely handle the crashing app
Fixed the button when expanded, cancel(if borrower)/see infos
Added tests

Patrick-Pataky reviewed (changes requested):
Awesome ! One last thing and than it's perfect:
![image](https://github.com/EPFLSWENT2024G1/partagix/assets/93245307/3915d2cf-4e63-4d0b-864e-3b23e7c51709)
- The default value for the email is `Please enter an email address`
- When you enter your mail in the Edit Account Screen, this default value should be considered as equivalent to `isEmpty()`

Suggestion: 
In `EditAccount.kt`:
- Add a new variable 
```
val defaultEmail = "Please enter an email addres"
```

- Line 206:
```suggestion
onValueChange = {value ->
    email = if (value.contains(defaultEmail.dropLast(1))) {
      ""
    } else {
      value
    }
    if (email == "") favorite[0] = false
  },
```
- Line 306:
```suggestion
enabled = favorite.contains(true) && email != defaultEmail,
```

LozanoCyrielle commented:
> Awesome ! One last thing and than it's perfect: ![image](https://private-user-images.githubusercontent.com/93245307/334208949-3915d2cf-4e63-4d0b-864e-3b23e7c51709.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY4NDQxMzAsIm5iZiI6MTcxNjg0MzgzMCwicGF0aCI6Ii85MzI0NTMwNy8zMzQyMDg5NDktMzkxNWQyY2YtNGU2My00ZDBiLTg2NGUtM2IyM2U3YzUxNzA5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI3VDIxMDM1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2YWFiODVmZWEzMWUxYmYxNmJlZWFiMzBhNzBlNDIwZWMyNDA2MDdjOWYxMzBiYjRkMjA0ZmQ2YWI4MjJhNzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vBbRUMGeSOorQ_UbTLSId62Pozp7-RbG9dMZBOEyQpU)
> 
> * The default value for the email is `Please enter an email address`
> * When you enter your mail in the Edit Account Screen, this default value should be considered as equivalent to `isEmpty()`
> 
> Suggestion: In `EditAccount.kt`:
> 
> * Add a new variable
> 
> ```
> val defaultEmail = "Please enter an email addres"
> ```
> 
> * Line 206:
> 
> ```
> onValueChange = {value ->
>     email = if (value.contains(defaultEmail.dropLast(1))) {
>       ""
>     } else {
>       value
>     }
>     if (email == "") favorite[0] = false
>   },
> ```
> 
> * Line 306:
> 
> ```
> enabled = favorite.contains(true) && email != defaultEmail,
> ```

Technically it's not the default value, it's a fail-safe. Because we already had users (without email address) when I started to code this I made a fail safe so the app would crash for user without any contact field filled.
With my new implementation when a user create an account his email is directly add as an email address so this should never happens. But you are right it's better to be prepared and have this, like that we don't have to be worried if something goes wrong with the email at first :))

Mass-14 reviewed (changes requested):
First of all, I love this new feature ! 

Here are some changes that would greatly improve the UI.
itemUi : 
- Remove the titles "Preferred contact:" and "Other contact:"
- Show only the means of contact that are correctly fulfilled (i.e. show the mail iff the format is correct)
- Put the preferred mean of contact in bold, and as the first of the list
- Put all means of contacts with the size `smallerFontSize`
- (optional) add an indent at each mean of contact (iff it looks cleaner)

EditAccount : 
- there should be only one favorite contact method
- add this property to all new Textfields `colors = OutlinedTextFieldDefaults.colors()` (will match the to first textfields)
- remove spacer line 248
- spacer line 289 to 8.dp
- add a spacer of 6.dp after line 313

LozanoCyrielle commented:
> First of all, I love this new feature !
> 
> Here are some changes that would greatly improve the UI. itemUi :
> 
> * Remove the titles "Preferred contact:" and "Other contact:"
> * Show only the means of contact that are correctly fulfilled (i.e. show the mail iff the format is correct)
> * Put the preferred mean of contact in bold, and as the first of the list
> * Put all means of contacts with the size `smallerFontSize`
> * (optional) add an indent at each mean of contact (iff it looks cleaner)
> 
> EditAccount :
> 
> * there should be only one favorite contact method
> * add this property to all new Textfields `colors = OutlinedTextFieldDefaults.colors()` (will match the to first textfields)
> * remove spacer line 248
> * spacer line 289 to 8.dp
> * add a spacer of 6.dp after line 313

Except for the only one favorite which I think should not be unique, I changed what you've said

Mass-14 reviewed (approved):
Very clean thank you !

Patrick-Pataky reviewed (approved):
Perfect !
